internal/hcs/system.go

```diff
@@ -326,13 +335,108 @@ func (computeSystem *System) Properties(ctx context.Context, types ...schema1.Pr
+			},
+			Processor: &hcsschema.ProcessorStats{
+				RuntimeKernel100ns: uint64(processorInfo.TotalKernelTime),
+				RuntimeUser100ns:   uint64(processorInfo.TotalUserTime),
+				TotalRuntime100ns:  uint64(processorInfo.TotalKernelTime + processorInfo.TotalUserTime),
+			},
+			Storage: &hcsschema.StorageStats{
+				ReadCountNormalized:  uint64(storageInfo.ReadStats.IoCount),
+				ReadSizeBytes:        storageInfo.ReadStats.TotalSize,
+				WriteCountNormalized: uint64(storageInfo.WriteStats.IoCount),
+				WriteSizeBytes:       storageInfo.WriteStats.TotalSize,
+			},
+		},
+	}, nil
+}
+
 // PropertiesV2 returns the requested container properties targeting a V2 schema container.
 func (computeSystem *System) PropertiesV2(ctx context.Context, types ...hcsschema.PropertyType) (*hcsschema.Properties, error) {
 	computeSystem.handleLock.RLock()
 	defer computeSystem.handleLock.RUnlock()
 
 	operation := "hcs::System::PropertiesV2"
 
+	// There's an optimization we can make here if the user is asking for Statistics only, and this is due to
+	// an inefficient way that HCS calculates the private working set total for a given container. HCS
+	// calls NtQuerySystemInformation with the class SystemProcessInformation which returns an array containing
+	// system information for *every* process running on the machine. They then grab the pids that are running in
+	// the container and filter down the entries in the array to only what's running in that silo and start tallying
+	// up the total. This doesn't work well as performance should get worse if more processess are running on the
+	// machine in general and not just in the container. All of the additional information besides the
+	// WorkingSetPrivateSize field is ignored as well which isn't great and is wasted work to fetch.
+	//
+	// HCS only let's you grab statistics in an all or nothing fashion, so we can't just grab the private
+	// working set ourselves and ask for everything else seperately. The optimization we can make here is
+	// to open the silo ourselves and do the same queries for the rest of the info, as well as calculating
+	// the private working set in a more efficient manner by:
+	//
+	// 1. Find the pids running in the silo
+	// 2. Get a process handle for every process (only need PROCESS_QUERY_LIMITED_INFORMATION access)
+	// 3. Call NtQueryInformationProcess on each process with the class ProcessVmCounters
+	// 4. Tally up the total using the field PrivateWorkingSetSize in VM_COUNTERS_EX2.
+	//
+	// The only caveat is the process invoking this code must be running as SYSTEM to open the silo,
+	// otherwise we'll fallback to asking HCS for the information.
+
+	// First check if we're only asking for Stats to see if we can apply the above optimization. We should
+	// only do this for containers and not UVMs.
+	if (len(types) == 1 && types[0] == hcsschema.PTStatistics) && computeSystem.typ == "container" {
+		properties, err := computeSystem.statisticsInProc(ctx)
```

Comment:
We should probably log this error since it will cause us to fall back to a slower path.

Reply:
Sure